### PR TITLE
Check exhaustivity of pattern matches with "if" guards and custom extractors (by default) and of unsealed types (by opt-in).

### LINKED ISSUE #5365: Match Exhaustiveness Testing Ignores Guard Statements
It seems that guarded match expressions are not tested for completeness at compile time. Or, more accurately, it seems that guarded case statements are optimistically considered complete. For example, the following code will compile without "warning: match is not exhaustive!", even though calling it with a value of 1 will result in a MatchError.

```
def foo(x: Option[Int]): Int = x match {
  case Some(n) if n % 2 == 0 => n
  case None => 0
}
```

The equivalent code in Ocaml does print a non-exhaustive match warning at compile time. 

```
let foo(x) = 
  match x with
  | Some n when n mod 2 = 0 -> n
  | None -> 0;;


Warning 8: this pattern-matching is not exhaustive.
Here is an example of a value that is not matched:
Some _
(However, some guarded clause may match this value.)
val foo : int option -> int = <fun>
```

Obviously the compiler can't decide whether or not arbitrary combinations of guards are complete, but I was surprised didn't treat guards pessimistically like Ocaml does. This actually bit us in production code, when a method like this:

```
def foo(x: Option[Int]): Int = x match {
  case Some(n) if n % 2 == 0 => n
  case _ => 0
}
```

was refactored to look like this:

```
def foo(x: Option[Int]): Int = x match {
  case Some(n) if n % 2 == 0 => n
  case None => 0
}
```

which resulted in runtime MatchErrors.

I understand that adding a warning that treats guarded match statements pessimistically could be annoying in cases when the combined guards actually are safe, but in these instances the warning could easily be dismissed with an annotation or a noop unguarded match. Why don't we give Scala users the option for added safety?

### LINKED ISSUE #8178: Case class with varargs defeats exhaustiveness matching
Including a case class with varargs in the constructor seems to prevent exhaustiveness checking from happening.
```scala
sealed trait Fails
case class VarArgs1(a: String*) extends Fails
case class FailsChild2(a: Seq[String]) extends Fails
object FailsTest {
  def matchOnVarArgsFirstFails(f: Fails) {
    f match {
      case VarArgs1(_) => ???
      // BUG: Without this line we should get a non-exhaustive match compiler error. 
      //case FailsChild2(_) => ???
    }
  }

  def matchOnSeqArgsFirstWorks(f: Fails) {
    f match {
      case FailsChild2(_) => ???
      // Without this line, the compiler reports a "match may not be exhaustive" error as expected.
      case VarArgs1(_) => ???
    }
  }
}

sealed trait Works
case class SeqArgs1(a: Seq[String]) extends Works
case class SeqArgs2(a: Seq[String]) extends Works
object WorksTest {
  def matcher(f: Works) {
    f match {
      case SeqArgs1(_) => ???
      // Without this line, the compiler reports a "match may not be exhaustive" error as expected.
      case SeqArgs2(_) => ???
    }
  }
}
```

### LINKED ISSUE #9232: unapplySeq silences all exhaustivity warnings
The following code compiles without any warning:
```scala
package test

final class Foo(val value: Int)
object Foo {
  def unapplySeq(foo: Foo): Some[Seq[Int]] = Some(List(foo.value))
  /*def unapply(foo: Foo): Some[Int] = Some(foo.value)*/
}

sealed trait Tree
case class Node1(foo: Foo) extends Tree
case class Node2() extends Tree

object Test {
  def transformTree(tree: Tree): Any = tree match {
    case Node1(Foo(1)) => ???
  }
}
```

But if you uncomment `unapply`, you get:
```scala
test2.scala:14: warning: match may not be exhaustive.
It would fail on the following inputs: Node1(_), Node2()
  def transformTree(tree: Tree): Any = tree match {
                                       ^
one warning found
```

It'd be nice to give the same warnings when only `unapplySeq` is used, for example with `List`.

### PR DESCRIPTION
**Pull request by dwijnand**
The accuracy of exhaustivity checking for pattern matches is improved. These changes intend to bring to light pattern matching code that might throw runtime `MatchError`s for unhandled cases, without any indication at compile time.

Most importantly the following no longer disable exhaustivity checking:

1. guards (fixing scala/bug#5365)
2. refutable custom extractors (irrefutable ones already didn't)
3. unsealed types (sealed types already didn't) by `-Xlint:strict-unsealed-patmat`

As before, it remains possible to disable checking of both exhaustivity and reachability by annotating an expression with `@unchecked`, such as `(foo: @unchecked) match { ... }`.

But you may also use `-Xnon-strict-patmat-analysis` to disable the "strict" pattern matching for guards (1) and custom extractors (2).

### Guards

Guards (`case .. if .. =>`) no longer disable the exhaustivity checker. Rather, the checker checks for exhaustivity using the "pessimistic" assumption that the guard may be `false`.

Thus, given the following example:

```scala
x match {
  case p if g1 => ...
  case p if g2 => ...
  case p if g3 => ...
}
```

The change decided is that, even if the user may know or be able to determine that one of `g1`, `g2` and `g3` will always be true (at least one), it's safer for the exhaustivity checker to not back off even if it means it will produce a warning here, requiring the user to suppress the warning by either:

1. dropping the final `if g3`; or
2. adding a default case (`case x => ...` or `case _ => ...`) with some body (perhaps throwing or logging)

### Custom Extractors

Refutable custom extractors no longer disable exhaustivity checking. (Irrefutable ones already didn't disable it).  It is “pessimistically” assumed that the extractor might not match.

An “irrefutable” custom extractor is one that returns `Some` rather than `Option`, such as:

```scala
object Length {
  def unapply(s: String): Some[Int] = Some(s.length)
}
```

So if you have custom extractors that are also irrefutable (that is they can't "not match") then you can resolve some exhaustivity warnings by making it return `Some[..]` rather than `Option[..]`.  As an example:

```scala
scala> object Length { def unapply(s: String): Option[Int] = Some(s.length) }
object Length

scala> "abc" match { case Length(n) => println(s"The length was $n.") }
       ^
       warning: match may not be exhaustive.
The length was 3.

scala> object Length { def unapply(s: String): Some[Int] = Some(s.length) }
object Length

scala> "abc" match { case Length(n) => println(s"The length was $n.") }
The length was 3.
```

### Unsealed types

Types that aren't sealed are also checked by the exhaustivity checker if the `-Xlint:strict-unsealed-patmat` option (or all of `-Xlint`) is enabled.  This makes pattern matching safer, as the compiler will now give a warning when it can't see that a pattern covers all the cases, rather than quietly disabling the check.

## Known Issues/Limitations

The counter-examples provided may not always be clear, particularly around `unapplySeq`.  For example it might say something like `Chain()` as an attempt to say that some values of type `Chain` aren't covered by the `Chain.unapplySeq` usage.

https://github.com/scala/bug/issues/12186 In some patterns, like type patterns, aliases of `case object`s (and possible `case class`es too) won't dealias and so won't register as exhausting the match, such as `case _: Nil.type =>` (with `Nil` being the val alias `Nil` in `Predef` rather than the `case object` in `scala.collection.immutable`).

Because these match cases are now considered the match analysis might now fail due to excessive recursion depth, for example:

```
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/src/compiler/scala/tools/nsc/typechecker/Typers.scala:5288:32: Exhaustivity analysis reached max recursion depth, not all missing cases are reported.
[error] (Please try with scalac -Ypatmat-exhaust-depth 160 or -Ypatmat-exhaust-depth off.)
[error]             val result = silent(_.makeAccessible(tree1, sym, pre, qual)) match {
[error]                                ^
[error] one error found
```

In addition to attempting to bump the threshold, you may find it easier to resolve this by breaking the match up into smaller parts (e.g. #9277).

We hope to resolve these soon... (🤞)

## Migration

The best-case scenario is that the matcher identifies holes in the match that were previously ignored, so these can be addressed by the programmer.

Where it is expected that some values will throw, you can get rid of the warning by explicitly defining a default case (`case x => ...` or `case _ => ...`), perhaps throwing a more specific exception with some more domain-specific information or recovering in some way. 

You might also choose the (sometimes) terser route of annotating some non-total pattern match by adding `@unchecked` as mentioned above.

## Provenance

This PR was originally:

* a continuation of @retronym’s and @sellout's work in #5617
* which @milessabin had made accessible as a part of Typelevel Scala 2.12.1 ([notes](https://github.com/typelevel/scala/blob/typelevel-readme/notes/2.12.1.md#exhaustivity-of-extractors-guards-and-unsealed-traits-pull5617-sellout))

Fixes scala/bug#5365
Fixes scala/bug#8178
Fixes scala/bug#9232

### CODE REVIEW

**Review by lrytz — APPROVED** [review]
LGTM!

I assume we'll explain all pattern matching changes in the 2.13.4 release notes. Nevertheless, it would be good to update the PR description here to summarize all changes that this brings.

**Comment by lrytz** [comment]
> summarize all changes that this brings

That might give us "higher level" reviews from more people without them having to look at the code or the individual commits.

**Comment by dwijnand** [comment]
> Nevertheless, it would be good to update the PR description here to summarize all changes that this brings.

Done. (ping @SethTisue for tweaks - all yours)

**Comment by sellout** [comment]
Thanks for picking this up, @dwijnand!

**Comment by SethTisue** [comment]
(I did an edit pass over the description.)

**Comment by SethTisue** [comment]
I'm delighted, and impressed, to see the matcher get so much smarter 🧠💪

But I have misgivings about this PR in its current form. I suggest some further thought/discussion about what the defaults should be and what mechanisms should be provided for altering those defaults.

In particular, I was surprised by the choice to make the optimistic rather than the pessimistic assumption about guards. Perhaps I just need to understand the reasoning behind that choice. But I had expected that exhaustivity should never be assumed if cannot actually be proven. And I think we want to move in the direction of exhaustive matches being considered the norm and requiring the user to be explicit about it when a match _isn't_ exhaustive, for greater safety overall.

Also (orthogonally), the new compiler flags aren't a good design IMO, because they're global. I think users will want to make this decision per-pattern-match, not per-project. How about an annotation? Might `@nowarn` be sufficient? And/or, are you aware of `@unchecked`'s behavior (https://stackoverflow.com/questions/42526314/how-to-suppress-the-match-may-not-be-exhaustive-warning-in-scala)?

And I don't understand why one of the new flags an `-Xlint:...` and one isn't? (Even if it makes sense, the PR description should briefly justify it.)

**Comment by dwijnand** [comment]
> In particular, I was surprised by the choice to make the optimistic rather than the pessimistic assumption about guards. Perhaps I just need to understand the reasoning behind that choice. But I had expected that exhaustivity should never be assumed if cannot actually be proven. And I think we want to move in the direction of exhaustive matches being considered the norm and requiring the user to be explicit about it when a match _isn't_ exhaustive, for greater safety overall.

That would absolutely be my preference too.  But this is far, far from the truth (sadly!!).

My experience maintaining MiMa leads me now to be more cautious when dealing with false positives (which are, regrettably, easy to come by in MiMa 😞), preferring an approach where a user opts into stricter-but-possibly-noisy checks (e.g. `mimaReportSignatureProblems := true` in MiMa) - mind you sbt-tpolecat et cetera kind of unintentionally undermines that, but 🤷‍♂️ ...

That said, where on the strict/noisy <-> quiet/risky spectrum we should pick for the various scenarios we may want to pick better choices.  I've been thinking about it and it's all about cases where the compiler warns but the user understands better the situation, and knows the warning is just wrong.  But I see a split in those cases:

(1) cases where the compiler doesn't know if the cases are exhaustive or not e.g.:

* (1a) complementary guards (`case Foo if n >= 0 => ... case Foo if n < 0 => ...`)
* (1b) complementary custom extractors (`case NonNegative(n) => ... case Negative(n) => ...`)

Here the knowledgeable user could be really annoyed by the noise and want everything to go back to not warning.

(2) cases where the compiler _does_ know the cases aren't exhaustive (even if they currently are for the user), e.g.:

* (2a) partial unappySeq extractors (`case List(x) => ... case List(x, _*) => ...`) "I know the list is always non-empty"  (List might be a poor example because there's some fudging going on with List)
* (2b) partial unsealed matches (`case n: Int => ... case s: String => ...`) "I know this Any is either ints or strings"

The latter I think we should consider being stricter by default, in priority/risk order (start with the less false positive-y):
1. case class unapplySeqs (see my last commit)
2. custom unapplySeqs
3. unsealed types

> Also (orthogonally), the new compiler flags aren't a good design IMO, because they're global. I think users will want to make this decision per-pattern-match, not per-project. How about an annotation? Might `@nowarn` be sufficient? And/or, are you aware of `@unchecked`'s behavior ([stackoverflow.com/questions/42526314/how-to-suppress-the-match-may-not-be-exhaustive-warning-in-scala](https://stackoverflow.com/questions/42526314/how-to-suppress-the-match-may-not-be-exhaustive-warning-in-scala))?

Somethings amiss: the flags are for more strictness, which I seems right to me: you dial up the strictness across the board (forSome board) and suppress on case-by-case bases (or, actually, across the board 😄).

> And I don't understand why one of the new flags an `-Xlint:...` and one isn't? (Even if it makes sense, the PR description should briefly justify it.)

Predates me and I didn't give it much thought.  Happy to make `-Xstrict-patmat-analysis` a lint if desired.

**Comment by smarter** [comment]
> Here the knowledgeable user could be really annoyed by the noise and want everything to go back to not warning.

I disagree. We write `val x = if (n >= 0) foo else bar` and not `val x = if (n >= 0) foo else if (n < 0) bar`, and if someone writes the latter then we tell them they're doing it wrong if they complain that the result type is `()` because there's no else branch. By the same reasoning, the compiler is correct to warn about a missing case even when someone writes complimentary guards or extractors: it's not a big deal to replace the complement by something that always matches, and I'd even argue it's clearer. This is how dotty behaves and I think it'd be great if scalac aligned itself to that.

**Comment by SethTisue** [comment]
I strongly agree with Guillaume about complementary guards and custom extractors. The examples you give are just bad code.

The only case I'm on the fence about whether to warn by default is unsealed types.

Re: the globalness of the flags, I'm suggesting the flags should not exist at all — they should be the defaults, and if someone doesn't like that default, they can `@nowarn` their way out of it. They're just warnings, not errors.

I'm curious what other repo watchers think.

**Comment by SethTisue** [comment]
@dwijnand I'm sort of curious if your thinking changed as a result of Guillaume and my comments, or if you just decided to go with the flow. (Perhaps idle curiosity, or, idk, perhaps understanding where we arrived at on the design thinking here will help us improve the PR description.)

I suggest we wait to merge (and to finalize the PR description) until @lrytz has weighed in, since he had approved the previous version.

**Review by SethTisue — APPROVED** [review]

**Comment by SethTisue** [comment]
small thought: `It would fail on the following input: _` seems like a common warning text. It feels... gnomic. Is there room for improvement there? (Should it say nothing in this case, since it isn't really adding information?)

**Comment by dwijnand** [comment]
> @dwijnand I'm sort of curious if your thinking changed as a result of Guillaume and my comments, or if you just decided to go with the flow. (Perhaps idle curiosity, or, idk, perhaps understanding where we arrived at on the design thinking here will help us improve the PR description.)

My heart/desire was (and still is) that exhaustivity were strict by default but I was on the fence.  Yours and Guillaume's certainty was enough for me to go that far, so we can assess if we like the change.

Thinking about it more today I suspect that where MiMa's were actually false positives (due to lack of scala pickle metadata), here the false positives due to complementary guards/extractors I believe are easier to grasp, and the lack of clear counter-examples are unfortunate but they aren't indicators of a false positive.

> small thought: `It would fail on the following input: _` seems like a common warning text. It feels... gnomic. Is there room for improvement there? (Should it say nothing in this case, since it isn't really adding information?)

```
t7623.scala:10: warning: match may not be exhaustive.
It would fail on the following input: _
  def f = C("") match { case C(s) => }
           ^
```

Yeah, I think it would look less silly if that just said "match may not be exhaustive".

**Comment by lrytz** [comment]
My thoughts after looking at the impact this has on tests, and after playing around for a while.

I'm happy to be asked for an `@unchecked` or `@nowarn` when some assumption is not represented in the types, like Dale's example "I know this Any is either Int or String".

I'm also happy to accept that the compiler bails on guards, buddy is a compiler after all, not a theorem prover. It's my code, and it's easy to make the match exhaustive by leaving the last guard away. As Guillaume says, it's like `if`-`else`.

However, I have an issue with spurious warnings for extractors. I think a main difference here is that extractors are typically not my code, or at least defined somewhere else than the match I'm writing. I fear that introducing spurious warnings will diminish the value of extractors for library designers.

For example, the following warns after this PR

```scala
scala> def m(s: Stream[Int]) = s match { case Stream.Empty => 0; case x #:: _ => x }
<console>:11: warning: match may not be exhaustive.
It would fail on the following input: Cons()
       def m(s: Stream[Int]) = s match { case Stream.Empty => 0; case x #:: _ => x }
                               ^
```

As a user, I'm not doing anything wrong here! Extractors enable a beautiful integration of libraries into the language. Pattern matching is a core language feature and the code looks really natural here. But the spurious warning really breaks the experience. It's not the same situation as with guards, where I can just leave the last one away. I need to call the extractor for it to do its job.

**Comment by smarter** [comment]
Yeah we have that warning on matching on LazyList/Stream, but to me it indicates that LazyList is just not a good fit for pattern-matching.

**Comment by dwijnand** [comment]
> For example, the following warns after this PR
> 
> ```scala
> scala> def m(s: Stream[Int]) = s match { case Stream.Empty => 0; case x #:: _ => x }
> <console>:11: warning: match may not be exhaustive.
> It would fail on the following input: Cons()
>        def m(s: Stream[Int]) = s match { case Stream.Empty => 0; case x #:: _ => x }
>                                ^
> ```
> 
> As a user, I'm not doing anything wrong here! Extractors enable a beautiful integration of libraries into the language. Pattern matching is a core language feature and the code looks really natural here. But the spurious warning really breaks the experience. It's not the same situation as with guards, where I can just leave the last one away. I need to call the extractor for it to do its job.

Wouldn't the fix be the same as with guards?

```scala
s match { case x #:: _ => x; case _ => 0; }
```

**Comment by lrytz** [comment]
The Stream is really just an example, I believe that the spurious warnings will affect library designers in ways that we don't want or don't expect. Extractors are a really powerful feature, they allow hooking into the syntax of pattern matching which is core to the language. I think the warnings are problematic.

**Comment by dwijnand** [comment]
If we go back, then I should remember to keep them strict behind `-Xstrict-patmat-analysis` (or as a lint).

Btw, just realised this contrast:

Seth:
> The only case I'm on the fence about whether to warn by default is unsealed types.

Lukas:
> I'm happy to be asked for an `@unchecked` or `@nowarn` when some assumption is not represented in the types, like Dale's example "I know this Any is either Int or String".

As implemented unsealed types require `-Xlint:strict-unsealed-patmat` for the checks.

**Comment by lrytz** [comment]
For example, let's say for some reason I want my library to be designed like this:

```scala
sealed trait T
final class A(val x: Int) extends T
object A {
  def unapply(a: A): Option[Int] = Option(a).map(_.x)
}
final class B(val x: Int, val y: Int) extends T
object B {
  def unapply(b: B): Option[(Int, Int)] = Option(b).map(b => (b.x, b.y))
}
```

Then for users

```scala
scala> def f(t: T) = t match { case A(x) => x; case B(x, y) => x + y }
<console>:14: warning: match may not be exhaustive.
It would fail on the following inputs: A(), B()
       def f(t: T) = t match { case A(x) => x; case B(x, y) => x + y }
                     ^
```

**Comment by smarter** [comment]
In that example you can make your extractors return `Some[Int]` instead of `Option[Int]` to let the compiler know that they always match and no warnings is needed (Dotty at least handles this correctly).

**Comment by lrytz** [comment]
OK, that makes sense, but it wouldn't work for a custom extrator like `#::`, for example.

**Comment by dwijnand** [comment]
I agree with Guillaume: that's a good counter-example of how the changes here could push for better extractor design by library authors.  If the extractor had always been refutable then it would push for better extractor usage by library users.

The problem with Stream.Empty and Stream.#:: is that they're both refutable, but together they're complementary, without the compiler knowing that.  The exact same thing _would_ be true of `Nil` and `List.unapply` if it weren't for the fact that the pattern matcher has code to handle that (actually 2 bits of code: a rewrite for the checking and some fudging for making counter-examples).  We could do the same for Stream/LazyList, if we wanted to...

**Comment by lrytz** [comment]
Thanks for being more positive, Guillaume & Dale!

I really would like to be convinced that the change is not too annoying (false warnings) or breaking, as I also support the direction this is going.

How eager is dotty with exhaustivity warnings, compared to this PR?

> `List.unapply` [...] pattern matcher has code to handle that

The fact that this was deemed necessary maybe says something about the user-friendliness of the default handling of extractors?

**Comment by smarter** [comment]
> How eager is dotty with exhaustivity warnings, compared to this PR?

I haven't checked in details but I think it's pretty similar now, here's how we determine whether we should check for exhaustivity/reachability: https://github.com/lampepfl/dotty/blob/101e620c86ebeed005905a5f401e985aeb3d8d80/compiler/src/dotty/tools/dotc/transform/patmat/Space.scala#L801-L815

> I really would like to be convinced that the change is not too annoying (false warnings) or breaking, as I also support the direction this is going.

One thing I wonder about is whether this is too disruptive for 2.12 (which this PR targets), I assume if you're the kind of person still running 2.12 you don't like updating your code so you probably won't like having to deal with new warnings too :).

**Comment by ijuma** [comment]
I'm looking forward to this, but I agree that 2.13 seems like a better target.

**Comment by lrytz** [comment]
I took another look. Questions and thoughts:

- The PR description is currently not in line with the implementation
  - `-Xstrict-patmat-analysis` is no longer there, the stricter checks are always on
  - Now there's `-Xno-unsealed-patmat-analysis` instead of `-Xlint:strict-unsealed-patmat`
- Should we provide flags to disable exhaustivity checking for guards and custom extractors? We're introducing new warnings in a minor release. I can imagine some users want to upgrade but not deal with new warnings right away.
- For a more gradual introduction, we could put the new warnings under `-Xlint`. They would be disabled by default, but enabled by default for people that use the default `-Xlint`. That would allow us to build confidence to enable them by default later.
- I also think we should target 2.13.x
- There are a bunch of `case x => throw new MatchError(x)` added in tests. That seems to be the kind of code one only has to write to make the compiler happy... Why didn't you use `@unchecked`? What is the recommendation for users?

**Comment by dwijnand** [comment]
> * The PR description is currently not in line with the implementation
>   
>   * `-Xstrict-patmat-analysis` is no longer there, the stricter checks are always on
>   * Now there's `-Xno-unsealed-patmat-analysis` instead of `-Xlint:strict-unsealed-patmat`

* [x] Yep, will update it.

> * Should we provide flags to disable exhaustivity checking for guards and custom extractors? We're introducing new warnings in a minor release. I can imagine some users want to upgrade but not deal with new warnings right away.

* [x] OK, I'll introduce `-Xnon-strict-patmat-analysis`

> * For a more gradual introduction, we could put the new warnings under `-Xlint`. They would be disabled by default, but enabled by default for people that use the default `-Xlint`. That would allow us to build confidence to enable them by default later.

* With the flag above and the move to 2.13 below, we'll leave this on by default, considering it an enhancement to the existing exhaustivity warning system, which previously failed to deliver in the presence of guards and custom extractors (and unsealed types)

> * I also think we should target 2.13.x

* [x] OK, will retarget

> * There are a bunch of `case x => throw new MatchError(x)` added in tests. That seems to be the kind of code one only has to write to make the compiler happy... Why didn't you use `@unchecked`? What is the recommendation for users?

This is because it's our regression suite.  I didn't want to impact the integrity of the tests, and `@unchecked` also disables reachability analysis.  (I also didn't want to spend hours reading back the git history and issue/PR comments of every impacted test file.)  So I made an educated guess, opting often to manually write the otherwise synthetic `throw MatchError` branch, to resolve warnings.  There are some cases that (from my reading) had to do with testing the synthetic throw branch, so I opted for `@unchecked` there.  I tended not to use `-Xno-unsealed-patmat-analysis` as that seemed _too_ implementation specific, and I wanted to keep the test cases forward-compatible, trying to mimic real user's code.

I actually think making the synthetic throw branch explicit is an ok change to make initially, for instance in an initial Scala version bump PR, with the intent of addressing those instances incrementally afterwards.  But I'll write down the recommendation for users in the releases notes (in the PR description).

**Comment by lrytz** [comment]
```
!!  433 - run/patmatnew.scala                       [output differs]
% diff /home/travis/build/scala/scala/test/files/run/patmatnew.check /home/travis/build/scala/scala/test/files/run/patmatnew-run.log
@@ -21,7 +21,7 @@ patmatnew.scala:175: warning: match may not be exhaustive.
     def doMatch2(xs: List[String]): List[String] = xs match {
                                                           ^
                                                     patmatnew.scala:261: warning: match may not be exhaustive.
-It would fail on the following inputs: (x: Any forSome x not in Test.TestSequence06.A), A((x: Any forSome x not in (1, Test.TestSequence06.A))), A(A((x: Any forSome x not in 1)))
+It would fail on the following inputs: (x: Any forSome x not in Test.TestSequence06.A), A((x: Any forSome x not in (1, Test.TestSequence06.A))), A((x: Any forSome x not in (1, Test.TestSequence06.A))), A(A((x: Any forSome x not in 1)))
     def doMatch(x: Any, bla: Int) = x match {
                                            ^
                                      patmatnew.scala:277: warning: match may not be exhaustive.
```

seems the `A((x: Any forSome x not in (1, Test.TestSequence06.A)))` counter example shows up twice now?

**Comment by SethTisue** [comment]
(remilestoned for 2.13.4, even though it hasn't been retargeted to 2.13.x yet)

**Review by SethTisue — APPROVED** [review]
I don't know about the implementation, but I'm satisfied designwise.

File: project/ScalaOptionParser.scala [hunk 1]
```diff
@@ -83,7 +83,7 @@ object ScalaOptionParser {
 
   // TODO retrieve these data programmatically, ala https://github.com/scala/scala-tool-support/blob/master/bash-completion/src/main/scala/BashCompletion.scala
   private def booleanSettingNames = List("-X", "-Xasync", "-Xcheckinit", "-Xdev", "-Xdisable-assertions", "-Xexperimental", "-Xfatal-warnings", "-Xlog-free-terms", "-Xlog-free-types", "-Xlog-implicit-conversions", "-Xlog-implicits", "-Xlog-reflective-calls",
-    "-Xno-forwarders", "-Xno-patmat-analysis", "-Xprint-pos", "-Xprint-types", "-Xprompt", "-Xresident", "-Xshow-phases", "-Xverify", "-Y",
+    "-Xno-forwarders", "-Xno-patmat-analysis", "-Xno-unsealed-patmat-analysis", "-Xprint-pos", "-Xprint-types", "-Xprompt", "-Xresident", "-Xshow-phases", "-Xverify", "-Y",
```
Comment (lrytz) [inline project/ScalaOptionParser.scala]: also add `-Xnon-strict-patmat-analysis`?

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala [hunk 2]
```diff
@@ -372,10 +388,11 @@ trait MatchApproximation extends TreeAndTypeAnalysis with ScalaLogic with MatchT
       }
 
       // special-case: interpret pattern `List()` as `Nil`
+      // as of 2.13, List.unapply returns an UnapplySeqWrapper (rather than a List)
       // TODO: make it more general List(1, 2) => 1 :: 2 :: Nil  -- not sure this is a good idea...
       private val rewriteListPattern: PartialFunction[TreeMaker, Prop] = {
         case p @ ExtractorTreeMaker(_, _, testedBinder)
-          if testedBinder.tpe.typeSymbol == ListClass && p.checkedLength == Some(0) =>
+          if testedBinder.tpe.typeSymbol == UnapplySeqWrapperClass && p.checkedLength == Some(0) =>
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala:395]: Doesn't this catch other uses of `SeqFactory.unapplySeq`, like 	`case Seq()`?

File: test/files/presentation/doc/doc.check [hunk 3]
```diff
@@ -0,0 +1,58 @@
+warning: 1 deprecation (since 2.13.0); re-run with -deprecation for details
+java.lang.AssertionError: assertion failed: doc/src is not a directory
```
Comment (lrytz) [inline test/files/presentation/doc/doc.check]: this stack trace in the check file seems accidental, or at least very prone to changes in the compiler that require updating the check file often

File: test/files/run/macro-vampire-false-warning/Macros_1.check [hunk 4]
```diff
@@ -0,0 +1,38 @@
+java.lang.ClassNotFoundException: Test
```
Comment (lrytz) [inline test/files/run/macro-vampire-false-warning/Macros_1.check]: also this stack trace

File: test/files/run/pure-warning-post-macro/Macro_1.check [hunk 5]
```diff
@@ -0,0 +1,38 @@
+java.lang.ClassNotFoundException: Test
```
Comment (lrytz) [inline test/files/run/pure-warning-post-macro/Macro_1.check]: stack trace

File: src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala (hunk 2 above)
Comment (dwijnand) [reply src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala:395]: Yeah.  Which is regrettable, but I couldn't convince myself that that was bad so I went with it...  Do you think it'll affect other things and/or have a better idea to fixing the regression?

**Comment by smarter** [comment]
For the record: When looking at this PR previously I missed that it turned on exhaustiveness checking for unsealed types by default (Dotty doesn't). I think this is a bigger and more controversial change than anything else in this PR (besides the new warnings, this could also have a performance impact since many more expressions will be checked by default). We discussed this at the Dotty meeting today (Dale was present) and Martin was pretty strongly against this being the default. We also touched on the fact that there's no way in source to determine which match expressions are checked or not which is unfortunate, and agreed this topic in general warranted further discussion.

**Comment by dwijnand** [comment]
Flipping the default in https://github.com/scala/scala/pull/9299.

File: src/library/scala/unchecked.scala [hunk 6]
```diff
@@ -15,7 +15,7 @@ package scala
 /** An annotation to designate that the annotated entity
  *  should not be considered for additional compiler checks.
  *  Specific applications include annotating the subject of
- *  a match expression to suppress exhaustiveness warnings, and
+ *  a match expression to suppress exhaustiveness and reachability warnings, and
```
Comment (lrytz) [inline src/library/scala/unchecked.scala:18]: Old PR, but this doesn't seem to be the case in Scala 2? You also mention it in https://github.com/scala/scala/pull/9140#issuecomment-699871332.

**Pull request merged**